modules/http/http.c

```diff
@@ -24,569 +24,656 @@
+    }
+
+  for (l = owner->headers; l; l = l->next)
+    headers = curl_slist_append(headers, l->data);
+
+  return headers;
 }
 
-void
-http_dd_set_method(LogDriver *d, const gchar *method)
+static void
+_add_message_to_batch(HTTPDestinationWorker *self, LogMessage *msg)
 {
-  HTTPDestinationDriver *self = (HTTPDestinationDriver *) d;
+  HTTPDestinationDriver *owner = (HTTPDestinationDriver *) self->super.owner;
 
-  if (g_ascii_strcasecmp(method, "POST") == 0)
-    self->method_type = METHOD_TYPE_POST;
-  else if (g_ascii_strcasecmp(method, "PUT") == 0)
-    self->method_type = METHOD_TYPE_PUT;
+  if (self->super.batch_size > 1)
+    {
+      g_string_append_len(self->request_body, owner->delimiter->str, owner->delimiter->len);
+    }
+  if (owner->body_template)
+    {
+      log_template_append_format(owner->body_template, msg, &owner->template_options, LTZ_SEND,
+                                 self->super.seq_num, NULL, self->request_body);
+    }
   else
     {
-      msg_warning("Unsupported method is set(Only POST and PUT are supported), default method POST will be used",
-                  evt_tag_str("method", method));
-      self->method_type = METHOD_TYPE_POST;
+      g_string_append(self->request_body, log_msg_get_value(msg, LM_V_MESSAGE, NULL));
     }
 }
 
-void
-http_dd_set_body(LogDriver *d, LogTemplate *body)
+static worker_insert_result_t
+_map_http_status_to_worker_status(HTTPDestinationWorker *self, glong http_code)
 {
-  HTTPDestinationDriver *self = (HTTPDestinationDriver *) d;
+  HTTPDestinationDriver *owner = (HTTPDestinationDriver *) self->super.owner;
+  worker_insert_result_t retval;
 
-  log_template_unref(self->body_template);
-  self->body_template = log_template_ref(body);
+  switch (http_code/100)
+    {
+    case 1:
+    case 2:
+      retval = WORKER_INSERT_RESULT_SUCCESS;
+      break;
+    case 3:
+      msg_notice("Server returned with a 3XX (redirect) status code, which was not handled by curl. "
+                 "Either accept-redirect() is set to no, or this status code is unknown. Trying again",
+                 evt_tag_str("url", owner->url),
+                 evt_tag_int("status_code", http_code),
+                 log_pipe_location_tag(&owner->super.super.super.super));
+      retval = WORKER_INSERT_RESULT_ERROR;
+      break;
+    case 4:
+      msg_notice("Server returned with a 4XX (client errors) status code, which means we are not "
+                 "authorized or the URL is not found. Trying again",
+                 evt_tag_str("url", owner->url),
+                 evt_tag_int("status_code", http_code),
+                 log_pipe_location_tag(&owner->super.super.super.super));
+      retval = WORKER_INSERT_RESULT_ERROR;
+      break;
+    default:
+      msg_error("Unknown HTTP response code",
+                evt_tag_str("url", owner->url),
+                evt_tag_int("status_code", http_code),
+                log_pipe_location_tag(&owner->super.super.super.super));
```

Comment:
We should set an error code in the default case too. Otherwise memory garbage is returned.

Reply:
the default was falling through to the 5xx case, so a default was set. but my followup patch makes that absolutely clear.